cli/emit.rs

```diff
@@ -126,6 +161,33 @@ impl<T: Cacher> EmitCache for T {
   }
 }
 
+/// Hash the set of options which can affect the result of a type check on a
+/// given module graph. This should be stored in the cached metadata for each
+/// root of the graph.
+///
+/// The config also includes the exact set of roots used. For example, if two
+/// type checks are performed and the second uses a subset of the same roots,
+/// it's possible that the first passes and the second fails because of missing
+/// global augmentations. But without including the roots in the config hash
+/// the second would pass the other preliminary checks and short-circuit.
+fn hash_check_config(
+  ts_config: &TsConfig,
+  mode: TypeCheckMode,
+  roots: &[(ModuleSpecifier, ModuleKind)],
+) -> String {
+  let mut input = vec![ts_config.as_bytes(), vec![mode as u8]];
```

Comment:
This vector could be preallocated to prevent resizing: `let mut input = Vec::with_capacity(2 + roots.len() * 2);`